moveit_ros/moveit_servo/src/utils/common.cpp

```diff
@@ -138,53 +146,115 @@ geometry_msgs::msg::Pose poseFromCartesianDelta(const Eigen::VectorXd& delta_x,
+      for (const auto& acc : state.accelerations)
+      {
+        point.accelerations.emplace_back(acc);
+      }
+    }
+    point.time_from_start = state.time_stamp - joint_trajectory.header.stamp;
+    joint_trajectory.points.emplace_back(point);
+  };
+
+  for (size_t i = 0; i < joint_cmd_rolling_window.size() - 1; ++i)
   {
-    point.positions.resize(num_joints);
-    point.velocities.resize(num_joints);
-    point.accelerations.resize(num_joints);
+    add_point(joint_trajectory, joint_cmd_rolling_window[i]);
   }
 
-  // Set the fields of trajectory point based on which fields are requested.
-  // Some controllers check that acceleration data is non-empty, even if accelerations are not used
-  // Send all zeros (joint_state.accelerations is a vector of all zeros).
+  return joint_trajectory;
+}
 
-  if (servo_params.publish_joint_positions)
+void updateSlidingWindow(KinematicState& next_joint_state, std::deque<KinematicState>& joint_cmd_rolling_window,
+                         double max_expected_latency)
+{
+  // remove old commands
+  const auto active_time_window = rclcpp::Duration::from_seconds(2 * max_expected_latency);
+  while (!joint_cmd_rolling_window.empty() &&
+         joint_cmd_rolling_window.front().time_stamp < (next_joint_state.time_stamp - active_time_window))
   {
-    for (size_t i = 0; i < num_joints; ++i)
-    {
-      point.positions[i] = joint_state.positions[i];
-    }
+    joint_cmd_rolling_window.pop_front();
   }
-  if (servo_params.publish_joint_velocities)
+
+  // remove command at end of window if timestamp is the same
+  if (!joint_cmd_rolling_window.empty() && next_joint_state.time_stamp == joint_cmd_rolling_window.back().time_stamp)
   {
-    for (size_t i = 0; i < num_joints; ++i)
-    {
-      point.velocities[i] = joint_state.velocities[i];
-    }
+    joint_cmd_rolling_window.pop_back();
   }
-  if (servo_params.publish_joint_accelerations)
+
+  // update velocity: the velocity has the potential to dramatically influence interpolation of splines and causes large
+  // overshooting. To alleviate this effect, the velocity will be set to zero if the motion changes direction,
+  // otherwise, it will calculate the forward and backward finite difference velocities and chose the minimum.
+  if (joint_cmd_rolling_window.size() >= 2)
   {
-    for (size_t i = 0; i < num_joints; ++i)
+    size_t num_points = joint_cmd_rolling_window.size();
+    auto& last_state = joint_cmd_rolling_window[num_points - 1];
+    auto& second_last_state = joint_cmd_rolling_window[num_points - 2];
+
+    Eigen::VectorXd direction_1 = second_last_state.positions - last_state.positions;
+    Eigen::VectorXd direction_2 = next_joint_state.positions - last_state.positions;
+    Eigen::VectorXd signs = (direction_1.array().sign() - direction_2.array().sign()).round();
+    for (long i = 0; i < last_state.velocities.size(); ++i)
     {
-      point.accelerations[i] = joint_state.accelerations[i];
+      if (abs(signs[i]) < 1.5)
```

Comment:
What's this 1.5? Is this basically checking that there are more than 1 sign flip? I think this value is fine, but may be worth a quick comment

Reply:
I added a comment explaining